File: java/src/test/java/com/bc/ur/URDecoderTest.java
package com.bc.ur;

import com.bc.ur.util.TestUtils;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

import java.util.Arrays;

import static com.bc.ur.URJni.UR_new_from_len_seed_string;
import static com.bc.ur.util.TestUtils.assertThrows;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;


@RunWith(JUnit4.class)
public class URDecoderTest {

    @Test
    public void testDecodeSinglePart() {
        UR expectedUR = UR_new_from_len_seed_string(50, "Wolf");
        String encoded = "ur:bytes/hdeymejtswhhylkepmykhhtsytsnoyoyaxaedsuttydmmhhpktpmsrjtgwdpfnsboxgwlbaawzuefywkdplrsrjynbvygabwjldapfcsdwkbrkch";
        UR ur = URDecoder.decode(encoded);
        assertEquals(expectedUR.getType(), ur.getType());
        assertTrue(Arrays.deepEquals(TestUtils.toTypedArray(expectedUR.getCbor()),
                                     TestUtils.toTypedArray(ur.getCbor())));
    }

    @Test
    public void testDecodeMultiParts() throws Exception {
        UR ur = UR_new_from_len_seed_string(32767, "Wolf");

        UREncoder refEncoder;
        URDecoder refDecoder;

        try (UREncoder encoder = new UREncoder(ur, 1000, 100, 10);
             URDecoder decoder = new URDecoder()) {
            refEncoder = encoder;
            refDecoder = decoder;

            long processPartCount = 0L;

            do {
                assertTrue(decoder.estimatedPercentComplete() < 1.0);

                String part = encoder.nextPart();
                decoder.receivePart(part);

                assertEquals(33L, decoder.expectedPartCount());
                assertEquals(++processPartCount, decoder.processedPartsCount());
            } while (!decoder.isComplete());

            assertTrue(decoder.isSuccess());
            assertFalse(decoder.isFailed());
            assertEquals(1.0, decoder.estimatedPercentComplete(), 1.0);

            // make sure resultUR return exact UR entered before
            UR resultUR = decoder.resultUR();
            assertEquals(ur.getType(), resultUR.getType());
            assertTrue(Arrays.deepEquals(TestUtils.toTypedArray(ur.getCbor()),
                                         TestUtils.toTypedArray(resultUR.getCbor())));

            // make sure getting resultError throw URException
            assertThrows("test failed due to checking resultError",
                         URException.class,
                         decoder::resultError);
        }

        // make sure encoder/decoder is closed
        assertTrue(refEncoder.isClosed());
        assertTrue(refDecoder.isClosed());

        assertThrows("test failed since encoder has not been disposed",
                     IllegalArgumentException.class,
                     refEncoder::nextPart);
        assertThrows("test failed since decoder has not been disposed",
                     IllegalArgumentException.class,
                     refDecoder::expectedType);
    }

    @Test
    public void testDecodeError() {
        String[] invalidData = new String[]{"",
                                            "ur:bytes/",
                                            "ur:ur:ur",
                                            "uf:bytes/hdeymejtswhhylkepmykhhtsytsnoyoyaxaedsuttydmmhhpktpmsrjtgwdpfnsboxgwlbaawzuefywkdplrsrjynbvygabwjldapfcsdwkbrkch"};
        for (String it : invalidData) {
            assertThrows("test failed due to " + it, URException.class, () -> URDecoder.decode(it));
        }
    }
}

File: android/app/src/androidTest/java/com/bc/ur/URDecoderTest.kt
package com.bc.ur

import androidx.test.ext.junit.runners.AndroidJUnit4
import com.bc.ur.URJni.UR_new_from_len_seed_string
import org.junit.Assert.*
import org.junit.Test
import org.junit.runner.RunWith


@RunWith(AndroidJUnit4::class)
class URDecoderTest {

    @Test
    fun testDecodeSinglePart() {
        val expectedUR = UR_new_from_len_seed_string(50, "Wolf")
        val encoded =
            "ur:bytes/hdeymejtswhhylkepmykhhtsytsnoyoyaxaedsuttydmmhhpktpmsrjtgwdpfnsboxgwlbaawzuefywkdplrsrjynbvygabwjldapfcsdwkbrkch"
        val ur = URDecoder.decode(encoded)
        assertEquals(expectedUR.type, ur.type)
        assertTrue(expectedUR.cbor.toTypedArray().contentDeepEquals(ur.cbor.toTypedArray()))
    }

    @Test
    fun testDecodeMultiParts() {
        val ur = UR_new_from_len_seed_string(32767, "Wolf")

        val encoder = UREncoder(ur, 1000, 100, 10)
        val decoder = URDecoder()

        encoder.use { e ->
            decoder.use { d ->

                var processPartCount = 0L
                do {
                    assertTrue(d.estimatedPercentComplete() < 1.0)

                    val part = e.nextPart()
                    d.receivePart(part)

                    assertEquals(33L, d.expectedPartCount())
                    assertEquals(++processPartCount, d.processedPartsCount())
                } while (!d.isComplete)

                assertTrue(d.isSuccess)
                assertFalse(d.isFailed)
                assertEquals(1.0, d.estimatedPercentComplete(), 1.0)


                // make sure resultUR return exact UR entered before
                val resultUR = d.resultUR()
                assertEquals(ur.type, resultUR.type)
                assertTrue(ur.cbor.toTypedArray().contentDeepEquals(resultUR.cbor.toTypedArray()))

                // make sure getting resultError throw URException
                assertThrows<URException>("test failed due to checking resultError") { d.resultError() }
            }
        }

        // make sure encoder/decoder is closed
        assertTrue(encoder.isClosed)
        assertTrue(decoder.isClosed)

        assertThrows<IllegalArgumentException>("test failed since encoder has not been disposed") { encoder.nextPart() }
        assertThrows<IllegalArgumentException>("test failed since decoder has not been disposed") { decoder.expectedType() }
    }

    @Test
    fun testDecodeError() {
        arrayOf(
            "",
            "ur:bytes/",
            "ur:ur:ur",
            "uf:bytes/hdeymejtswhhylkepmykhhtsytsnoyoyaxaedsuttydmmhhpktpmsrjtgwdpfnsboxgwlbaawzuefywkdplrsrjynbvygabwjldapfcsdwkbrkch"
        ).forEach {
            assertThrows<URException>("test failed due to $it") { URDecoder.decode(it) }
        }

    }
}